### PR TITLE
scale ingress controller to the number of kubernetes-workers

### PR DESCRIPTION
**Pull request by wwwtyro**
@chuckbutler @mbruzek

Scales the ingress controller replica count to match the number of kubernetes-worker units.

We see a couple of issues with this PR:

1. There's a potential race condition wherein the controller replica count can be set to a value less than the number of kubernetes-worker units. It seems unlikely to occur, but we may need to look into placing this logic elsewhere or running it at some interval instead of only once.

2. The scaling appears to occur serially, at least how we're invoking it here. In our three-node microbot test, each nginx ingress pod took 11 minutes to reach a `Running` state, and it took 33 minutes before we were able to reach microbot because the last nginx ingress pod was the one selected by microbot to serve as its proxy.

### CODE REVIEW

**Comment by chuckbutler** [comment]
i want to run an additional scenario with this code where we over-provision the replicas and see what behavior is introduced into kubernetes.

consider the following where you're adding replicas, a user juju removes-unit, and the cluster is left trying to converge to a 50 node kubernetes-worker farm, where there's only actually 36 deployed units.

If we pass then, i'm +1 to merging this as is.

**Comment by mbruzek** [comment]
I was not sure how this would put one pod on each node. @Cynerva convinced me that each one with the port would distribute to each node.

File: cluster/juju/layers/kubernetes-worker/reactive/kubernetes_worker.py [hunk 1]
```diff
@@ -211,6 +212,13 @@ def render_and_launch_ingress(kube_dns):
         hookenv.close_port(80)
         hookenv.close_port(443)
 
+def scale_ingress_controller():
```
Comment (mbruzek) [inline cluster/juju/layers/kubernetes-worker/reactive/kubernetes_worker.py]: Docstring needed.

File: cluster/juju/layers/kubernetes-worker/reactive/kubernetes_worker.py [hunk 2]
```diff
@@ -211,6 +212,13 @@ def render_and_launch_ingress(kube_dns):
...
+    kubectl = ['kubectl', '--kubeconfig=/srv/kubernetes/config']
+    command = kubectl + ['get', 'nodes', '-o', 'name']
+    output = check_output(command, shell=False)
+    count = len(output.splitlines())
+    command = kubectl + ['scale', '--replicas=%d' % count, 'rc/nginx-ingress-controller']
```
Comment (mbruzek) [inline cluster/juju/layers/kubernetes-worker/reactive/kubernetes_worker.py]: Will this scale down?  We should take a look.

**Comment by wwwtyro** [comment]
@mbruzek 

Added the docstring.

Tested scaling up. Hit the race condition predicted earlier, I think. Changed things so that it checks every update and scales as necessary. Should have eventual consistency this way. In testing, scaling up worked smoothly.

Scaling down has another issue. Nodes are not removed when a worker unit is removed, so we don't have an accurate count of nodes (and kubernetes thinks there's more nodes than there are). Fix for this seems straightforward at first glance, so I'm going to try to slap together a PR for it real quick. Scaling down worked smoothly with this code when I manually removed a node.

**Comment by chuckbutler** [comment]
Hey rye, on the worker node stop hook execute a delete node on itself.

Kubectl delete node <<address>>

This should help keep the node consistency problem to a minimum.

**Comment by Cynerva** [comment]
> Hey rye, on the worker node stop hook execute a delete node on itself.

Addressed in a separate PR: #54

File: cluster/juju/layers/kubernetes-worker/reactive/kubernetes_worker.py [hunk 3]
```diff
@@ -194,6 +193,16 @@ def render_and_launch_ingress(kube_dns):
         hookenv.close_port(80)
         hookenv.close_port(443)
 
+@when('kubernetes-worker.ingress.available')
+def scale_ingress_controller():
+    ''' Scale the number of ingress controller replicas to match the number of nodes. '''
+    kubectl = ['kubectl', '--kubeconfig=/srv/kubernetes/config']
```
Comment (chuckbutler) [inline cluster/juju/layers/kubernetes-worker/reactive/kubernetes_worker.py]: This should be using the kubectl command gate declared on line 336.

**Comment by mbruzek** [comment]
+1 this seems to work.

**Pull request merged**